### ISSUE TITLE
python black formatter don't work with quarto: format cell

### ISSUE THREAD

**Issue by jjallaire** [quarto-dev/quarto#111]

**Comment by cderv** [comment]
For reference, related Posit Community discussion https://community.rstudio.com/t/format-python-code-in-quarto-r-markdown-files/159452

and other slightly related topic: 
* Issue in black to support markdown code chunk: https://github.com/psf/black/issues/294
* I opened issue the side project that build on black for this purpose (Markdown docs),to support Qmd file: https://github.com/adamchainz/blacken-docs/issues/222

**Comment by jjallaire** [comment]
I've gotten this to work with the following commit: https://github.com/quarto-dev/quarto/commit/c9d57dbdb4c6f65a7d94ddf9ecd0d660c41fd223

However, I don't think the result is going to be all that great (and I'm not sure what to do with it).

When we call a formatter for a language on a qmd document, we actually create a "virtual" document of the specified language. When we do this we include only the lines of code from the appropriate language cells. Note that other lines in the file _cannot_ be blank lines (otherwise the formatting engine will collapse them). So instead we feed it lines with comments. So for example the following `.qmd` document:

````markdown

```{python}
def function(x, y):
    return x + y
```

````

Is transformed to the following `.py` document for formatting:

```python
#
#
def function(x, y):
    return x + y
#
#
```

Unfortunately when you run the black formatter on this document it inserts blank lines in between the code and comment lines:

```python
#
#
def function(x, y):
    return x + y


#
#
```

If you run this on the `.qmd` you'll see the blank lines show up:

````markdown

```{python}
def function(x, y):
    return x + y


```

````

Note that this occurs as well when using black to format an entire qmd document (each cell gets trailing newlines).

Perhaps there is a way to configure black to not do this?

**Comment by cderv** [comment]
> Perhaps there is a way to configure black to not do this?

I did not find any configuration for this. 
I believe the two new lines are from [empty lines rules](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#empty-lines)

> It will also insert proper spacing before and after function definitions. It’s one line before and after inner functions and **two lines before and after module-level functions and classes**. Black will not put empty lines between function/class definitions and standalone comments that immediately precede the given function/class.

I tried using in [balck playground](https://black.vercel.app/) the [special deactivation comments](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#code-style)` # fmt: on/off` for bloc that we replace by `#`, but the new lines are inserted as part of the definition reformating 
````
def function(x, y):
    return x + y
````

I guess the definition of black gives us hint that won't be possible to change "The uncompromising Python code formatter."

**Comment by grst** [comment]
As far as I can tell the "vscode jupyter" extension sends each cell individually to the formatter when executing "Format Document". Would that be an option? 

When I format a minimal notebook with two cells it generates the following log in `Output -> Black Formatter`: 
```
2023-08-07 13:34:19.355 [info] [Trace - 1:34:19 PM] Sending request 'textDocument/formatting - (11)'.
2023-08-07 13:34:19.355 [info] [Trace - 1:34:19 PM] Sending request 'textDocument/formatting - (12)'.
2023-08-07 13:34:19.358 [info] [Trace - 1:34:19 PM] Received notification 'window/logMessage'.
2023-08-07 13:34:19.358 [info] /home/<path redacted>/bin/black --stdin-filename /home/<path redacted>/test.py -
2023-08-07 13:34:19.358 [info] [Trace - 1:34:19 PM] Received notification 'window/logMessage'.
2023-08-07 13:34:19.358 [info] CWD Server: /home/<path redacted>/
2023-08-07 13:34:19.917 [info] [Trace - 1:34:19 PM] Received notification 'window/logMessage'.
2023-08-07 13:34:19.917 [info] reformatted /home/<path redacted>/test.py

All done! ✨ 🍰 ✨
1 file reformatted.

2023-08-07 13:34:19.917 [info] [Trace - 1:34:19 PM] Received response 'textDocument/formatting - (11)' in 562ms.
2023-08-07 13:34:19.919 [info] [Trace - 1:34:19 PM] Received notification 'window/logMessage'.
2023-08-07 13:34:19.919 [info] /home/<path redacted>/bin/black --stdin-filename /home//test.py -
2023-08-07 13:34:19.919 [info] [Trace - 1:34:19 PM] Received notification 'window/logMessage'.
2023-08-07 13:34:19.919 [info] CWD Server: /home/<path redacted>/
2023-08-07 13:34:20.179 [info] [Trace - 1:34:20 PM] Received notification 'window/logMessage'.
2023-08-07 13:34:20.179 [info] reformatted /home/<path redacted>/test.py

All done! ✨ 🍰 ✨
1 file reformatted.

2023-08-07 13:34:20.179 [info] [Trace - 1:34:20 PM] Received response 'textDocument/formatting - (12)' in 824ms.
```

**Comment by jjallaire** [comment]
We could definitely consider this! Right now we send a comment line for all non-cell code (so that the offsets of the returned formatting commands match up). These comment lines are then getting collapsed by the black formatter.

So its not quite as simple as just sending a block at a time, but with some clever offsetting we could make this work.

**Comment by jjallaire** [comment]
It can auto-format within the current cell but not the entire document (for the reason described above). We will likely find a way to make this work for entire documents but not in the immediate future.

**Comment by agnz** [comment]
> It can auto-format within the current cell but not the entire document (for the reason described above). We will likely find a way to make this work for entire documents but not in the immediate future.

Thanks for a quick reply. Is there something special I need to do to set it up? Currently if I run Format Cell, I get the following error in Black output log

```
  File "<unknown>", line 81
    ```
    ^
SyntaxError: invalid syntax
```

**Comment by jjallaire** [comment]
Interesting, I'm not seeing that error however I am seeing this warning:

```python
2023-11-27 21:44:55.726 [info] [Warn  - 9:44:55 PM] Black does not support range formatting. Formatting entire document.
2023-11-27 21:44:55.726 [info] /usr/local/bin/python3 -m black --stdin-filename /Users/jjallaire/Desktop/.vdoc.py -
2023-11-27 21:44:55.726 [info] CWD formatter: /Users/jjallaire/Desktop
2023-11-27 21:44:55.727 [info] All done! ✨ 🍰 ✨
```

So part of the problem is that we are aren't able to do range formatting with black, and that causes things to not quite work the way they are supposed to. Will investigate further but won't be able to do this in the immediate future.

**Comment by jjallaire** [comment]
I did manage to figure out how to make the black formatter behave correctly with the Format Cell command. Further, Format on Save is also hooked up to invoke Format Cell so if you save as you are editing cells you'll have a decent experience. This is available in v1.108.0 of the extension which has been published to the marketplace.

**Comment by agnz** [comment]
This is great! Thank you for putting some time into this!
One hiccup - it seems to be really unhappy with quarto directives at the start of a cell, here's an example:
```
``` {python}
# | label: fig-pytorch-finetuning-all1
# | fig-cap: Improvement of the PyTorch model after fine-tuning using RS training set

# Average improvement from baseline to best for the pytorch model
plot_pr_curve(('pytorch', False,'half'), ['framework', 'cropped','resolution'], data=df, style='model', label=False)
```
```
On save, this gets formatted into:
```
``` {python}
# | label: fig-pytorch-finetuning-all1
# | fig-cap: Improvement of the PyTorch model after fine-tuning using RS training set

    "" ""
   """" ""
   
   ""
   ,

```
```

However, if I insert one blank line right at the top of the cell, it gets formatted correctly into
```
``` {python}

# | label: fig-pytorch-finetuning-all1
# | fig-cap: Improvement of the PyTorch model after fine-tuning using RS training set

# Average improvement from baseline to best for the pytorch model
plot_pr_curve(
    ("pytorch", False, "half"),
    ["framework", "cropped", "resolution"],
    data=df,
    style="model",
    label=False,
)
```
```
Not sure if it is an issue with black or the extension

**Comment by jjallaire** [comment]
Hmm..I am not actually able to reproduce that! Not sure what is going on.

**Comment by jjallaire** [comment]
Thanks very much for reporting this and for the detailed repro! This issue has been fixed and the fix has been published to the marketplace in version v1.109.0
